examples/ex26.cpp

```diff
@@ -0,0 +1,243 @@
+//               simple finite element discretization of the Laplace problem
+//               -Delta u = 1 with homogeneous Dirichlet boundary conditions
+//               as in example 1. It highlights on the creation of a hierarchy
+//               of discretization spaces with partial assembly and the
+//               construction of an efficient multigrid preconditioner for the
+//               iterative solver.
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+
+class DiffusionMultigrid : public Multigrid
+{
+private:
+   ConstantCoefficient one;
+
+public:
+   /// Constructor for a diffusion multigrid for a given FiniteElementSpaceHierarchy.
+   /// Uses Chebyshev accelerated smoothing.
+   DiffusionMultigrid(FiniteElementSpaceHierarchy& fespaces, Array<int>& ess_bdr)
+      : Multigrid(fespaces), one(1.0)
+   {
+      ConstructCoarseOperatorAndSolver(fespaces.GetFESpaceAtLevel(0), ess_bdr);
+
+      for (int level = 1; level < fespaces.GetNumLevels(); ++level)
+      {
+         ConstructOperatorAndSmoother(fespaces.GetFESpaceAtLevel(level), ess_bdr);
+      }
+   }
+
+private:
+   void ConstructBilinearForm(FiniteElementSpace& fespace, Array<int>& ess_bdr)
+   {
+      BilinearForm* form = new BilinearForm(&fespace);
+      form->SetAssemblyLevel(AssemblyLevel::PARTIAL);
+      form->AddDomainIntegrator(new DiffusionIntegrator(one));
+      form->Assemble();
+      bfs.Append(form);
+
+      essentialTrueDofs.Append(new Array<int>());
+      fespace.GetEssentialTrueDofs(ess_bdr, *essentialTrueDofs.Last());
+   }
+
+   void ConstructCoarseOperatorAndSolver(FiniteElementSpace& coarse_fespace,
+                                         Array<int>& ess_bdr)
+   {
+      ConstructBilinearForm(coarse_fespace, ess_bdr);
+
+      OperatorPtr opr;
+      opr.SetType(Operator::ANY_TYPE);
+      bfs.Last()->FormSystemMatrix(*essentialTrueDofs.Last(), opr);
+      opr.SetOperatorOwner(false);
+
+      CGSolver* pcg = new CGSolver();
+      pcg->SetPrintLevel(-1);
+      pcg->SetMaxIter(200);
+      pcg->SetRelTol(sqrt(1e-4));
+      pcg->SetAbsTol(0.0);
+      pcg->SetOperator(*opr.Ptr());
+
+      AddLevel(opr.Ptr(), pcg, true, true);
+   }
+
+   void ConstructOperatorAndSmoother(FiniteElementSpace& fespace,
+                                     Array<int>& ess_bdr)
+   {
+      ConstructBilinearForm(fespace, ess_bdr);
+
+      OperatorPtr opr;
+      opr.SetType(Operator::ANY_TYPE);
+      bfs.Last()->FormSystemMatrix(*essentialTrueDofs.Last(), opr);
+      opr.SetOperatorOwner(false);
+
+      Vector diag(fespace.GetTrueVSize());
+      bfs.Last()->AssembleDiagonal(diag);
+
+      Solver* smoother = new OperatorChebyshevSmoother(opr.Ptr(), diag,
+                                                       *essentialTrueDofs.Last(), 2);
+      AddLevel(opr.Ptr(), smoother, true, true);
+   }
+};
+
+int main(int argc, char *argv[])
```

Comment:
Two lines between the class and `main()`?

Reply:
I added a newlines here and in ex26p